src/Extensions/CompanionServer.cpp

```diff
@@ -18,148 +18,143 @@
 #include "Extensions/CompanionServer.hpp"
 #include "Core/EventLogger.hpp"
 #include "Core/MessageLogger.hpp"
+#include "qhttpfwd.hpp"
+#include "qhttpserver.hpp"
+#include "qhttpserverconnection.hpp"
+#include "qhttpserverrequest.hpp"
+#include "qhttpserverresponse.hpp"
 #include <QJsonArray>
 #include <QJsonObject>
-#include <QTcpServer>
-#include <QTcpSocket>
+
+#define USER_INFO(x)                                                                                                   \
+    if (log)                                                                                                           \
+        log->info("Companion", x);
+
+#define USER_WARN(x)                                                                                                   \
+    if (log)                                                                                                           \
+        log->warn("Companion", x);
+
+#define USER_ERR(x)                                                                                                    \
+    if (log)                                                                                                           \
+        log->error("Companion", x);
 
 namespace Extensions
 {
-CompanionServer::CompanionServer(int port)
+CompanionServer::CompanionServer(int port, QObject *parent) : QObject(parent)
 {
-    updatePort(port);
+    startListeningOn(port);
 }
 
 void CompanionServer::setMessageLogger(MessageLogger *log)
 {
     this->log = log;
 }
 
+bool CompanionServer::startListeningOn(int port)
+{
+    if (server)
+        delete server;
+    server = new qhttp::server::QHttpServer(this);
+    server->listen(QString::number(port), [this](qhttp::server::QHttpRequest *req, qhttp::server::QHttpResponse *res) {
+        LOG_INFO("\n--> " << req->methodString() << " : " << qPrintable(req->url().toString().toUtf8()));
+        req->headers().forEach(
+            [](auto iter) { LOG_INFO(iter.key().constData() << " : " << iter.value().constData()); });
+
+        const QString methodType = req->methodString();
+        const bool isJson = req->headers().keyHasValue("content-type", "application/json");
+        req->collectData();
+
+        req->onEnd([=] {
```

Comment:
Why not capture by reference?

Reply:
I think they are local variables and will be destroyed at the end of `CompanionServer::startListeningOn`, which results in undefined behavior.